### PR TITLE
Creating `MSnExperiment` from an ordinary list of `Spectrum` and other changes

### PR DESCRIPTION
**Pull request by jotsetung**
This (rather lengthy - sorry for that) pull request covers:
- Add `MSnExperiment` constructor function to enable creation of an `MSnExperiment` from a `list` of `Spectrum` objects. This is crucial if we want to replace `Spectra` (issue #436).
- Add `featureData` and `featureData<-` methods for `MSnExperiment` (`spectraData` and `spectraData<-` link to these). This is also crucial for #436 because we need a easy way to add arbitrary metadata columns to the `MSnExperiment`. Also, this closes issue #439 and is related to #438.
- Add a `backendUpdateMetadata` method to update spectrum metadata in the backend. This is currently not required for `BackendMzR` and `BackendHdf5`, but is needed for `BackendMemory` and eventually other backends that save not only the *mz* and *intensity* values. With that we can not call `spectraData(mse)$polarity <- 1L` and the polarity will be updated in all spectra of the `BackendMemory`.
- Expanded `spectraData` checking in `MSnExperiment` that we don't get nasty surprises with `spectraData<-`.
- Add a `sampleData<-` method to enable updating the `sampleData` in `MSnExperiment`.

### CODE REVIEW

**Comment by sgibb** [comment]
While reading and thinking about this two questions arise in my head:

1. Do we really need a `MSnExperiment` constructor for a `list` of `Spectrum` objects and a `readMSnExperiment` for `mzR` related files? (and if we want to support SQL databases we would need an `readMSnExperimentFromSQL` function).
Wouldn't it be useful to create a `MSnExperiment(..., backend)` constructor. That uses the backend `backend` to import the data, e.g `msexp <- MSnExperiment(files, backend=BackendMzR())`. Subsequently we could use a different backend to work with the data `msexp <- setBackend(msexp, backend=BackendHdf5())`. This would allow use to use `msexp <- MSnExperiment(listOfSpectra, backend=BackendMemory())` or after the user wrote his own sql backend:  `msexp <- setBackend(MSnExperiment(sqlitefile, backend=BackendSql()), backend=BackendHdf5())`.

2. Is `backendUpdateMetadata` really necessary? Wouldn't it be enough to update just the `spectraData`/`featureData` of the `MSnExperiment`. If `backendReadSpectra`/`backendWriteSpectra` is called the modified `spectraData` are given as argument and the backend could update the metadata on its own if necessary (as we doing it now with the `fromFile` slot in `backendSubset,BackendMemory`, however the `.spectrum_set_header` function is really useful).
If we have a file based backend that stores its metadata in files too each `featureData(msnexperiment)$fileIdx <- 1` (or something usefull) call will rewrite all metadata which slows down the process. In contrast the `backendReadSpectra/backendWriteSpectra` approach will only change the metadata if requested and will apply all metadata changes in one go (also a more or less lazy processing than `backendUpdateMetadata`).

**Comment by jotsetung** [comment]
Regarding 1:
I would not mix using the `backend` to store the data and `backend` to be used as a data importer. I would rather make a constructor `MSnExperiment(x, ..., backend)` that rather switches on `x`. Because we could also want to create a `MSnExperiment` from a `list` of spectra and use the `BackendHdf5`. For the SQL backend I have in mind I have anyway a different approach. I have a `CompoundDb` class that interfaces the database. There are a lot of other methods to retrieve data and annotations from the database and there is this one `spectra,CompoundDb` function that returns an `MSnExperiment` with a `BackendCompoundDb` backend that takes care of retrieving the data.

Regarding 2:
I think yes. Not for `BackendMzR` or `BackendHdf5`, but for `BackendMemory` or any other future backend that stores also spectrum metadata. If we don't have `backendUpdateMetadata` every modification of the spectrum metadata will lead to data corruption as the metadata in `@spectraData` and within each `Spectrum` is different.
Yes, it slows down some calls where you change metadata columns, but it is not too bad - and the current `MSnExp` did in fact exactly the same - so, `BackendMemory` will not be slower than `MSnExp`. And yes, we could implement this in `backendReadSpectra`, i.e. updating the spectrum metadata for each spectrum that is returned with the current values in `@spectraData`, but IMHO this would be even slower, given that getting data is a more frequent task than setting data. If there is no really strong argument against it I would really like to keep it this way.

File: R/MSnExperiment-class.R [hunk 1]
```diff
@@ -243,6 +319,72 @@ setClass(
...
+#' @param x `DataFrame`
+#'
+#' @param y `DataFrame`
+#'
+#' @return `DataFrame` with the merged columns.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.merge_featureData <- function(x, y) {
```
[the remark this answers is not in the document]
Comment (jotsetung) [reply R/MSnExperiment-class.R]: and what about `combine_featureData` or `join_featureData`? `cbind` means to me *adding another column to a matrix*, while this function does only add columns if they are not already there. In other words, from `cbind` I would expect that the resulting table has `ncol` equal to `ncol(x) + ncol(y)`.

[the remark this answers is not in the document]
Comment (sgibb) [reply R/MSnExperiment-class.R]: `combine_featureData` is great.

[the remark this answers is not in the document]
Comment (jotsetung) [reply R/MSnExperiment-class.R]: I'll fix that. It will be in PR #443

**Pull request merged**